### PR TITLE
chore: get push queries working with windowed primitive keys

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

Fixes: https://github.com/confluentinc/ksql/issues/4095

The main issue here was that things like `DataSourceNode` and `StepSchemaResolver` were incorrectly copying `ROWKEY` into the value schema as-is for windowed keys. This was out of line with what the transformer for windowed keys was doing in `SourceBuilder`. The transformer was converting the windowed key into a `STRING` in the form: `<key> : Window:{start=<start>, end=<end>}`. Hence the SqlType of the `ROWKEY` in the value schema should be `STRING` for windowed sources.

NOTE: this means that the type of `ROWKEY` can not be different in the key and value schema, e.g. a windowed source with a `BIGINT` key will have `ROWKEY BIGINT KEY` as the key schema, but the copy of `ROWKEY` added to the source schema as it passes through the Streams topology will be `ROWKEY STRING` to reflect that fact its stored as a STRING, e.g. `"11 :  Window:{start=149532753000, end=-}"`.

This disconnect between the Sql types of `ROWKEY` in the key and value schema should be temporary. Once we support truly structured keys this will go.

This commit makes this change.

* `LogicalSchema.withMetaAndKeyColsInValue` now takes a `boolean` to indicate if the source is windowed or not.
* As the type of `ROWKEY` can not be different in the key and value schem `ExpressionTypeManager` needs to get the type from the value schema for push queries (or actually anything not a pull query).

### Testing done 

Suitable QTT test cases and unit tests added.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by big-andy-coates** [comment]
Waiting on https://github.com/confluentinc/ksql/pull/4331 before I merge to avoid merge hell on that big PR for Rohan.

File: ksql-functional-tests/src/test/resources/query-validation-tests/key-schemas.json [hunk 1]
```diff
@@ -350,6 +350,89 @@
...
+            "type": "table",
+            "keyFormat": {"format": "KAFKA", "windowType": "TUMBLING", "windowSize": 30000},
+            "keyField": null,
+            "schema": "`ROWKEY` BIGINT KEY, `ID` BIGINT, `MAX` BIGINT"
+          }
+        ]
+      }
+    },
+    {
+      "name": "windowed stream explicit non-STRING ROWKEY udf",
```
Comment (rodesai) [inline ksql-functional-tests/src/test/resources/query-validation-tests/key-schemas.json]: this is not a windowed stream right?

**Comment by big-andy-coates** [comment]
> The code here LGTM in that it enables projecting the ROWKEY. Stepping back a bit though, wouldn't it make more sense (until we actually support a structured windowed key) for `ROWKEY` to always have type `VARCHAR` for windowed sources? We can't really use ROWKEY as anything other than a STRING right, e.g what happens if I try to use it in an expression (e.g.`SELECT ROWKEY / 1000` for an integer key)?

I did start down that route. Unfortunately, we still need the type of `ROWKEY` in the KEY schema to be the actual underlying type, e.g. `INT`, not always `STRING`, because KSQL needs to know how to serialize the data.  Remember, while `ROWKEY` may be displayed in the form `{ someKey : Window{start=10, end=11}`, that's not how its been serialized, even when it was only STRING keys.  It would be serialized as the binary form of the two Java longs `10` and `11`, followed by the Java serialized bytes for `someKey`.  Where the underlying key is not something other than `STRING` I think we still want the resultant data to be correct, e.g. an `INTEGER` type should be serialized as a Java serialized `int`, not the string representation of it.

TBH, the whole `ROWKEY` thing is a bit of a mess when it comes to windowed - on that we can agree.  The `ROWKEY` in the key is NOT the same as `ROWKEY` in the value.  This was the case _before_ primitive key support.

This will get cleaned up with structured key support.

**Comment by rodesai** [comment]
> serialized as the binary form

Ah right, forgot about this bit. Fair enough.

File: ksql-functional-tests/src/test/resources/query-validation-tests/key-schemas.json (hunk 1 above)
Comment (big-andy-coates) [reply ksql-functional-tests/src/test/resources/query-validation-tests/key-schemas.json]: It's a good point, well made ;)

**Pull request merged**